### ISSUE TITLE
[question] can i use ppo in multi gpu mode?

### ISSUE THREAD

**Issue by wujiren** [DLR-RM/stable-baselines3#158]
like supervised learning model, if i want use multi gpus in training a algorithm such as ppo, could i add some DataParallel code in OnPolicyAlgorithm as below? would the code work?

```python
class OnPolicyAlgorithm(BaseAlgorithm):
    def __init__(some_params):
        super(OnPolicyAlgorithm, self).__init__(some_params)
        
       "some codes"

        if _init_setup_model:
            self._setup_model()

    def _setup_model(self) -> None:
       "some codes"
        policy = self.policy_class(
            self.observation_space,
            self.action_space,
            self.lr_schedule,
            use_sde=self.use_sde,
            device=self.device,
            **self.policy_kwargs  # pytype:disable=not-instantiable
        )

        # add pytorch multi gpu code
        self.policy = torch.nn.DataParallel(policy)
        self.policy = self.policy.cuda()

```

**Comment by Miffyli** [comment]
We do not really offer custom tech support like this, and I can not right off the bat say if this would work. However I highly doubt this will speed up training as the main bottleneck usually are environments and synchronous execution of them, not the network computations.

**Comment by wujiren** [comment]
thanks for your guidance. I am going to check the time consume further.
at now, i read the ppo log, it seems the algorithm used more time in gpu calculation.  some of the log output is like below:
-------------------------------
| time/                           |               |
|    fps                           | 1164      |
|    iterations                 | 1            |
|    time_elapsed           | 450       |
|    total_timesteps       | 524288 |
-------------------------------
-----------------------------------------
| time/                               |                |
|    fps                               | 415         |
|    iterations                    | 2            |
|    time_elapsed              | 2524      |
|    total_timesteps          | 1048576|
| train/                              |                |
|    approx_kl                    | 0.00529|
|    clip_fraction               | 0             |
|    clip_range                  | 0.2          |
|    entropy_loss              | -0.675    |
|    explained_variance   | 0.843      |
|    learning_rate             | 0.001       |
|    loss                            | 0.0167      |
|    n_updates                 | 366         |
|    policy_gradient_loss | -1.96e-05   |
|    value_loss                  | 0.0945      |
-----------------------------------------

as many paper mentioned, ppo benefits from a bigger sample size, so i used a relatively bigger sample size of 524k. it seems the gpu consumed 3 times more than cpus. actually, i just used 1/3 of the cpus and memory resources and just put 1/10 of the data in the env。so the first thing i thought is using more gpu to accelerate the process since i can use up to 8 gpus.
 
i konw it is out of your job, and thanks for the reply at the weekend.i have searched some other rl resources, they do not have multi-gpu too. i am very appreciate that stable-baselines3 is so convenient for algorithm learning,testing and re-development, so i am trying to use sb3 to implement a multi gpu rl algorithm. due to my unprofessional knowledge, i may stack in a lot of troubles, so it will be appreciate if you can give me more guidance or other papers or documents about multi gpu rl training.
thanks again for your helps.

**Comment by Miffyli** [comment]
To use multiple machines/GPUs fully things get more complicated than just parallelizing backpropagation etc. See [this](https://arxiv.org/abs/1602.01783) and [this](https://arxiv.org/pdf/1802.01561.pdf) paper as main examples. Such asynchronous modifications are currently beyond SB3 but _might_ be considered in further versions.

**Comment by Karlheinzniebuhr** [comment]
> thanks for your guidance. I am going to check the time consume further. at now, i read the ppo log, it seems the algorithm used more time in gpu calculation. some of the log output is like below:
> 
> | time/ | | | fps | 1164 | | iterations | 1 | | time_elapsed | 450 | | total_timesteps | 524288 |
> 
> | time/ | | | fps | 415 | | iterations | 2 | | time_elapsed | 2524 | | total_timesteps | 1048576| | train/ | | | approx_kl | 0.00529| | clip_fraction | 0 | | clip_range | 0.2 | | entropy_loss | -0.675 | | explained_variance | 0.843 | | learning_rate | 0.001 | | loss | 0.0167 | | n_updates | 366 | | policy_gradient_loss | -1.96e-05 | | value_loss | 0.0945 |
> 
> as many paper mentioned, ppo benefits from a bigger sample size, so i used a relatively bigger sample size of 524k. it seems the gpu consumed 3 times more than cpus. actually, i just used 1/3 of the cpus and memory resources and just put 1/10 of the data in the env。so the first thing i thought is using more gpu to accelerate the process since i can use up to 8 gpus.
> 
> i konw it is out of your job, and thanks for the reply at the weekend.i have searched some other rl resources, they do not have multi-gpu too. i am very appreciate that stable-baselines3 is so convenient for algorithm learning,testing and re-development, so i am trying to use sb3 to implement a multi gpu rl algorithm. due to my unprofessional knowledge, i may stack in a lot of troubles, so it will be appreciate if you can give me more guidance or other papers or documents about multi gpu rl training. thanks again for your helps.

Did you find any other solution? My code is considerably slower with CUDA vs CPU :(
